base/dcp_mutation_feed.go

```diff
@@ -0,0 +1,103 @@
+package base
+
+import (
+	"context"
+	"errors"
+	"expvar"
+	"fmt"
+
+	"github.com/couchbase/gocbcore/v10"
+	sgbucket "github.com/couchbase/sg-bucket"
+)
+
+// getHighSeqMetadata returns metadata to feed into a DCP client based on the last sequence numbers stored in memory
+func getHighSeqMetadata(bucket Bucket) ([]DCPMetadata, error) {
+	store, ok := AsCouchbaseStore(bucket)
+	if !ok {
+		return nil, errors.New("DCP Client requires bucket to be CouchbaseStore")
+	}
+
+	numVbuckets, err := store.GetMaxVbno()
+	if err != nil {
+		return nil, fmt.Errorf("Unable to determine maxVbNo when creating DCP client: %w", err)
+	}
+
+	vbUUIDs, highSeqNos, statsErr := store.GetStatsVbSeqno(numVbuckets, true)
+	if statsErr != nil {
+		return nil, fmt.Errorf("Unable to obtain high seqnos for one-shot DCP feed: %w", statsErr)
+	}
+
+	metadata := make([]DCPMetadata, numVbuckets)
+	for vbNo := uint16(0); vbNo < numVbuckets; vbNo++ {
+		metadata[vbNo].VbUUID = gocbcore.VbUUID(vbUUIDs[vbNo])
+		metadata[vbNo].FailoverEntries = make([]gocbcore.FailoverEntry, 0)
+		metadata[vbNo].StartSeqNo = gocbcore.SeqNo(highSeqNos[vbNo])
+		metadata[vbNo].EndSeqNo = gocbcore.SeqNo(uint64(0xFFFFFFFFFFFFFFFF))
+		metadata[vbNo].SnapStartSeqNo = gocbcore.SeqNo(highSeqNos[vbNo])
+		metadata[vbNo].SnapEndSeqNo = gocbcore.SeqNo(highSeqNos[vbNo])
+	}
+	return metadata, nil
+}
+
+// StartGOCB2DCPFeed starts a DCP Feed.
+func StartGOCB2DCPFeed(bucket Bucket, spec BucketSpec, args sgbucket.FeedArguments, callback sgbucket.FeedEventCallbackFunc, dbStats *expvar.Map) error {
```

Comment:
It doesn't look like we're actually using BucketSpec here (could get spec.BucketName from bucket.Name()).  

If possible I think it would be preferable to pass in the parameters we actually need (including the collection set in future) instead of passing spec around.  In general I think the use case for BucketSpec should be 'I need to establish a new connection to the server from scratch'.

Reply:
This is a somewhat challenging problem to address right now as per offline discussion with @bbrks. We need the `Bucket` type in `DCPClient` if we are using persistent metadata. Moreover `Bucket` is actually a `Collection` which is conceptually wrong for a `DCPClient`, since it will be acting for multiple collections in a future code review, and we need actually pass something that can persist metadata for multiple collections, which oddly `Bucket` can not do.

I think we _do_ want to change this API, but not right now.